File: lecture06/02_subsets/subsets.c
/* subsets.c */

#include <stdio.h>

/* Functions */

int divisible_subsets(int n) {
    int count = 0;

    for (int bitset = 0; bitset < (1<<10); bitset++) {
    	int total = 0;
    	for (int i = 0; i < 10; i++) {
    	    total += (bitset & (1<<i)) ? i : 0;
	}
	count += (total % n == 0);
    }

    return count;
}

/* Main execution */

int main(int argc, char *argv[]) {
    int n;

    while (scanf("%d", &n) != EOF) {
    	printf("%d\n", divisible_subsets(n));
    }

    return 0;
}

File: lecture06/02_subsets/subsets.py
#!/usr/bin/env python3

import sys

# Function

def divisible_subsets(n):
    count = 0
    for bitset in range(1<<10):
        total = sum(i for i in range(10) if bitset & (1<<i))
        if total % n == 0:
            count += 1
    return count

def main():
    for n in map(int, sys.stdin):
        print(divisible_subsets(n))

# Main execution

if __name__ == '__main__':
    main()
